astropy/io/fits/hdu/image.py

```diff
@@ -135,9 +147,31 @@ def __init__(self, data=None, header=None, do_not_scale_image_data=False,
                 self._data_needs_rescale = True
             return
         else:
+            # Setting data will set set _bitpix, _bzero, and _bscale to the
+            # appropriate BITPIX for the data, and always sets _bzero=0 and
+            # _bscale=1.
             self.data = data
             self.update_header()
 
+            # Check again for BITPIX/BSCALE/BZERO in case they changed when the
+            # data was assigned. This can happen, for example, if the input
+            # data is an unsigned int numpy array.
+            self._bitpix = self._header.get('BITPIX')
+
+            # Do not provide default values for BZERO and BSCALE here because
+            # the keywords will have been deleted in the header if appropriate
+            # after scaling. We do not want to put them back in if they
+            # should not be there.
+            self._bzero = self._header.get('BZERO')
+            self._bscale = self._header.get('BSCALE')
```

Comment:
> Forgive me if all of the above is obvious to you already.

Not at all -- what I was missing was some sense of what the internal behavior *should be*, this is very helpful.